### PR TITLE
[SPARK-18430] [SQL] Fixed Exception Messages when Hitting an Invocation Exception of Function Lookup

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
When the exception is an invocation exception during function lookup, we return a useless/confusing error message:

For example, 
```Scala
df.selectExpr("concat_ws()")
```
Below is the error message we got:
```
null; line 1 pos 0
org.apache.spark.sql.AnalysisException: null; line 1 pos 0
```

To get the meaningful error message, we need to get the cause. The fix is exactly the same as what we did in https://github.com/apache/spark/pull/12136. After the fix, the message we got is the exception issued in the constuctor of function implementation:
```
requirement failed: concat_ws requires at least one argument.; line 1 pos 0
org.apache.spark.sql.AnalysisException: requirement failed: concat_ws requires at least one argument.; line 1 pos 0
```

### How was this patch tested?
Added test cases.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 1]
```diff
@@ -446,7 +446,10 @@ object FunctionRegistry {
         // If there is an apply method that accepts Seq[Expression], use that one.
         Try(varargCtor.get.newInstance(expressions).asInstanceOf[Expression]) match {
           case Success(e) => e
-          case Failure(e) => throw new AnalysisException(e.getMessage)
+          case Failure(e) =>
+            // the exception is an invocation exception. To get a meaningful message, we need the
+            // cause.
+            throw new AnalysisException(e.getCause.getMessage)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Above is [the same fix](https://github.com/gatorsmile/spark/blob/a43acced66fa15228b6c32176bb9f28dacb6f024/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala#L465-L468) in https://github.com/apache/spark/pull/12136.

**Comment by rxin** [comment]
Does this apply to all functions? If yes, I'd put a test in SQLQueryTestSuite

**Comment by gatorsmile** [comment]
Sure, let me do it.

**Comment by rxin** [comment]
Merging in master/branch-2.1. Thanks.

**Comment by rxin** [comment]
Can you submit a backport for branch-2.0?

**Comment by gatorsmile** [comment]
Sure, will do it. Thanks!

**Pull request closed without merging**